### ISSUE TITLE
Is it possible to use pre-built binaries via local filepath instead of a mirror?

### ISSUE THREAD

**Issue by olsn** [Automattic/node-canvas#1502]
I was wondering if it is somehow possible to use a locally placed pre-compiled binary instead, so that the process won't have to download it during the install.
By company policy our build-server is disconnected from the internet and everything must be able to be built without an internet-connection. The node-modules are served by a local nexus in the intranet.
Because of this we are not able to use the pre-built binaries nor built it from the source, since the built seems to require an internet-connection as well. (at least it is trying to download something from a cloudflare-ip 104.20.23.46)

We verified that everything works fine as long as there is access to the internet. Both, the pre-built-binaries as well as building from sources. But as soon as there is no internet:
- the pre-built download of course not
- but also building from source does not work in that case

For node-sass e.g. we where able to solve this via an env-var that points to the pre-compiled file on the local filesystem.

## Your Environment
As a build-container we use a docker-container (based on node-stretch-10).

We do know about the option to setup an own mirror to serve the prebuilt binaries, however this would mean quite a bit of overhead since this would require us to apply for firewall-exemptions etc.:
And to use a local fileserver in the build-container, just for the purpose of the install seems kind of "wrong" to me.

**Comment by zbjornson** [comment]
The only way I know how to change the download source is to fork node-canvas and change package.json according to the configuration info here: https://github.com/mapbox/node-pre-gyp#configuring. You might find some other solutions searching for things like "node-pre-gyp override download" or something.

(node-sass doesn't use node-pre-gyp; they have some other solution for prebuilds.)

**Comment by LinusU** [comment]
> since the built seems to require an internet-connection as well

Hmmm, this shouldn't be necessary 🤔 we should look into this...

**Comment by chearon** [comment]
> since the built seems to require an internet-connection as well

That's probably `node-gyp` downloading the headers for the current node, check out https://github.com/nodejs/node-gyp/issues/1020#issuecomment-244718826.

There's no equivalent for `node-pre-gyp` though. There's the [binary_host_mirror argument](https://github.com/mapbox/node-pre-gyp#download-binary-files-from-a-mirror) but you already said that won't work for you. Maybe some post install script to unzip the binaries from a local folder? I'd recommend looking at the first option to build from source though.

**Comment by olsn** [comment]
Thanks for all the recommendations!
postinstall with unzip does not work, since the install(or `npm ci`) aborts if one package fails .. yes we could force it, but that's not a very good solution, since e.g. we'd potentially miss other failures.

What we now did: We installed node-canvas globally and ship it with the build-container (yes, we are allowed to build the docker build-container with internet on a different system and load it into the private intranet repository ... don't ask ... it's company policy) and then we have a `"postinstall": "npm link canvas"` - it's not pretty, but it works and does not cause any negative side-effects.
